src/java/org/apache/cassandra/config/DataRateSpec.java

```diff
@@ -27,99 +27,88 @@
 
 import org.apache.cassandra.exceptions.ConfigurationException;
 
+import static org.apache.cassandra.config.DataRateSpec.DataRateUnit.BYTES_PER_SECOND;
+import static org.apache.cassandra.config.DataRateSpec.DataRateUnit.MEBIBYTES_PER_SECOND;
+
 /**
  * Represents a data rate type used for cassandra configuration. It supports the opportunity for the users to be able to
  * add units to the confiuration parameter value. (CASSANDRA-15234)
  */
-public final class DataRateSpec
+public abstract class DataRateSpec
 {
     /**
      * The Regexp used to parse the rate provided as String in cassandra.yaml.
      */
-    private static final Pattern BIT_RATE_UNITS_PATTERN = Pattern.compile("^(\\d+)(MiB/s|KiB/s|B/s)$");
+    private static final Pattern UNITS_PATTERN = Pattern.compile("^(\\d+)(MiB/s|KiB/s|B/s)$");
 
     private final double quantity;
 
     private final DataRateUnit unit;
 
-    public DataRateSpec(String value)
+    private DataRateSpec(String value)
     {
         //parse the string field value
-        Matcher matcher = BIT_RATE_UNITS_PATTERN.matcher(value);
+        Matcher matcher = UNITS_PATTERN.matcher(value);
 
         if (!matcher.find())
-            throw new ConfigurationException("Invalid bit rate: " + value + " Accepted units: MiB/s, KiB/s, B/s where " +
+            throw new ConfigurationException("Invalid data rate: " + value + " Accepted units: MiB/s, KiB/s, B/s where " +
                                              "case matters and " + "only non-negative values are valid");
 
-        quantity = Long.parseLong(matcher.group(1));
+        quantity = (double) Long.parseLong(matcher.group(1));
         unit = DataRateUnit.fromSymbol(matcher.group(2));
     }
 
-    DataRateSpec(double quantity, DataRateUnit unit)
+    private DataRateSpec(String value, DataRateUnit minUnit, long max)
     {
-        if (quantity < 0)
-            throw new ConfigurationException("Invalid bit rate: value must be non-negative");
+        this (value);
 
-        if (quantity > Long.MAX_VALUE)
-            throw new NumberFormatException("Invalid bit rate: value must be between 0 and Long.MAX_VALUE = 9223372036854775807");
+        validateQuantity(value, quantity(), unit(), minUnit, max);
```

Comment:
Why there isn't a `validateMinUnit` here?

Reply:
Because DataRate doesn’t have the sense of Min Unit. We can go down in units because internally we have it stored in double, even if for users it is only ling number allowed to provide. But we have the bound of particular unit though. The rate limiter those properties use is using already double and because of the conversion for backward compatibility to/from megabits per second this was done like that.